third_party/accessibility/ax/platform/ax_platform_node_textrangeprovider_win.cc

```diff
@@ -0,0 +1,1701 @@
+  HRESULT hr = MoveEndpointByUnitImpl(TextPatternRangeEndpoint_Start, unit,
+                                      count, &start_units_moved);
+
+  bool succeeded_move = SUCCEEDED(hr) && start_units_moved != 0;
+  if (succeeded_move) {
+    SetEnd(start()->Clone());
+    if (!is_degenerate_range) {
+      bool forwards = count > 0;
+      if (forwards && start()->AtEndOfDocument()) {
+        // The start is at the end of the document, so move the start backward
+        // by one text unit to expand the text range from the degenerate range
+        // state.
+        int current_start_units_moved = 0;
+        hr = MoveEndpointByUnitImpl(TextPatternRangeEndpoint_Start, unit, -1,
+                                    &current_start_units_moved);
+        start_units_moved -= 1;
+        succeeded_move = SUCCEEDED(hr) && current_start_units_moved == -1 &&
+                         start_units_moved > 0;
+      } else {
+        // The start is not at the end of the document, so move the endpoint
+        // forward by one text unit to expand the text range from the degenerate
+        // state.
+        int end_units_moved = 0;
+        hr = MoveEndpointByUnitImpl(TextPatternRangeEndpoint_End, unit, 1,
+                                    &end_units_moved);
+        succeeded_move = SUCCEEDED(hr) && end_units_moved == 1;
+      }
+
+      // Because Windows ATs behave undesirably when the start and end endpoints
+      // are not in the same anchor (for character and word navigation), make
+      // sure to bring back the end endpoint to the end of the start's anchor.
+      if (start()->anchor_id() != end()->anchor_id() &&
+          (unit == TextUnit_Character || unit == TextUnit_Word)) {
+        ExpandToEnclosingUnitImpl(unit);
+      }
+    }
+  }
+
+  if (!succeeded_move) {
+    SetStart(std::move(start_backup));
+    SetEnd(std::move(end_backup));
+    start_units_moved = 0;
+    if (!SUCCEEDED(hr))
+      return hr;
+  }
+
+  *units_moved = start_units_moved;
+  return S_OK;
+}
+
+HRESULT AXPlatformNodeTextRangeProviderWin::MoveEndpointByUnit(
+    TextPatternRangeEndpoint endpoint,
+    TextUnit unit,
+    int count,
+    int* units_moved) {
+  return MoveEndpointByUnitImpl(endpoint, unit, count, units_moved);
+}
+
+HRESULT AXPlatformNodeTextRangeProviderWin::MoveEndpointByUnitImpl(
+    TextPatternRangeEndpoint endpoint,
+    TextUnit unit,
+    int count,
+    int* units_moved) {
+  UIA_VALIDATE_TEXTRANGEPROVIDER_CALL_1_OUT(units_moved);
+
+  // Per MSDN, MoveEndpointByUnit with zero count has no effect.
+  if (count == 0) {
+    *units_moved = 0;
+    return S_OK;
+  }
+
+  bool is_start_endpoint = endpoint == TextPatternRangeEndpoint_Start;
+  AXPositionInstance position_to_move =
+      is_start_endpoint ? start()->Clone() : end()->Clone();
+
+  AXPositionInstance new_position;
+  switch (unit) {
+    case TextUnit_Character:
+      new_position =
+          MoveEndpointByCharacter(position_to_move, count, units_moved);
+      break;
```

Comment:
Why was `TextUnit_Format` removed? Should there be a TODO?

Reply:
I don't know, I will see if it can be added

Reply:
I see. We do not actually have `kFormatStart` and `kFormatEnd` in our `TextBoundary` enum.

Reply:
Should we create an issue and leave a TODO then? This seems like a useful thing: https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src/WinAuto/uiauto-uiautomationtextunits.md#format